core/sail/lucene-api/src/main/java/org/eclipse/rdf4j/sail/lucene/LuceneSail.java

```diff
@@ -551,6 +551,15 @@ public void shutDown() {
 						}
 						statements.add(vf.createStatement(r, p, o, c));
 					}
+
+					// make sure to index statements for last resource
+					if (current != null) {
+						if (logger.isDebugEnabled()) {
+							logger.debug("reindexing resource " + current);
+						}
+						// commit
+						luceneIndex.addDocuments(current, statements);
```

Comment:
Just a a general note, the whole code above should actually be using a peeking iterator. It's iterating over all the statements in the database and grouping by subject. It's doing it in a streaming fashion, which makes me assume that the iterator it has against the database is sorted by subject (which would take a massive amount of memory!). And it needs to check the next statement to see if it's got a different subject so it can add all the statements it's gathered for that subject to the lucene index as a document. Is that correct? 

I'm not proposing that we need to actually rewrite the code.

Could you check that !statements.empty() as well, so we don't add the last document twice?

Reply:
> Just a a general note, the whole code above should actually be using a peeking iterator. It's iterating over all the statements in the database and grouping by subject. It's doing it in a streaming fashion, which makes me assume that the iterator it has against the database is sorted by subject (which would take a massive amount of memory!). And it needs to check the next statement to see if it's got a different subject so it can add all the statements it's gathered for that subject to the lucene index as a document. Is that correct?

Yeah I think that sums it up. This code is pretty old, and I agree that there are likely far more efficient ways to trigger a reindexing. 

> I'm not proposing that we need to actually rewrite the code.

Worth logging an issue for? The problem is that none of us have LuceneSail maintenance/improvements high on our priority list.

> Could you check that !statements.empty() as well, so we don't add the last document twice?

Good point. I'll add a check for that.